proposals/0000-jsem.rst

```diff
@@ -0,0 +1,332 @@
+here.
+
+We also believe that GHC invocations should not be too eager to release
+resources, so we propose a minimum delay between acquiring a resource and
+releasing a resource. This allows us to prioritise completing a single large
+package before moving on to other work. The exact mechanism through which this
+is achieved is also left unspecified in this proposal.
+
+Examples
+--------
+
+Let us explain how we envision ``cabal`` handle the following build plan, with
+8 capabilities.
+
+.. image:: jsem_modules_plain.svg
+  :width: 600
+  :alt: Same dependency graph as before: a big unit at the top and bottom,
+        and many single-module units in the middle.
+
+1. To start, ``cabal`` would create a semaphore ``⟨sem⟩`` with 8 available tokens.
+
+2. Next, we compile the ``Bot`` unit, which is a large unit, with many modules,
+   which sits at the bottom of the dependency graph and must thus be compiled
+   before anything else.
+
+   a. ``cabal`` acquires one token from the semaphore and spawns one
+      ``ghc --make -jsem ⟨sem⟩`` invocation.
+   b. This invocation of ``ghc`` notices it has a lot of work to do (many modules
+      to compile from the ``Bot`` unit), so it requests more resources from the
+      semaphore, obtaining the remaining 7 tokens.
+   c. ``ghc`` finishes compiling the ``Bot`` unit, releasing the 7 tokens it
+      acquired.
+   d. ``cabal`` notices the ``ghc`` subprocess has terminated, and releases
+      the final (8th) token to the semaphore.
+
+3. After that, we move to compiling the middle units.
+
+   a. ``cabal`` will acquire tokens from the semaphore and spawn
+      ``ghc --make -jsem ⟨sem⟩`` invocations.
+   b. These ``ghc`` invocations won't query for more resources from the
+      semaphore, so ``cabal`` will manage running 8 concurrent ``ghc`` processes,
+      spawning new ones as previous ones terminate.
+
+4. Once all the middle units are compiled, ``cabal`` will move on to compiling
+   the top unit, which will proceed as in (2) with a single
+   ``ghc --make -jsem ⟨sem⟩`` invocation compiling 8 modules in parallel.
+
+5. Once all ``ghc`` processes have terminated, we are done, and ``cabal``
+   destroys ``⟨sem⟩``.
+
+In this situation, ``cabal`` is the jobserver: it manages the semaphore and
+spawns ``ghc`` subprocesses. The ``ghc`` subprocesses are jobclients, and they
+communicate by use of the semaphore.
+
+We can also imagine ``cabal`` or ``stack`` being jobclients, if there is a
+parent jobserver. For example, this would happen if ``nix`` manages the
+semaphore, with ``nix`` spawning ``cabal`` subprocesses, and those ``cabal``
+subprocesses spawning ``ghc`` subprocesses in turn.
+
+Effects and Interactions
+------------------------
+
+1. The ``-jsem`` and ``-j`` flags override eachother.
+2. The implementation of multiple home units can provide an alternative way
+   to achieve the same goals: compilation is achieved using a single GHC
+   invocation, so no coordination is necessary.
```

Comment:
This could do with more discussion I think, probably in its own subsection. What are the trade-offs between having a single `--make` session using MHU vs using `-jsem`? Do we have reason one would be easier to implement in `cabal` or `stack` or `nix`?

Reply:
I've fleshed it out into a subsection based off today's discussion.